File: C++/binarySearch.cpp
/*
    Contributed By : Shubham
    Date Modified: 01, October , 2022
    github id : https://github.com/im-shubham
    linkedinId : https://www.linkedin.com/in/shubhamvig/
*/

#include<bits/stdc++.h>
using namespace std;

// BinarySearch implementation
int binarySearch(vector<int> &v, int element_toFind){
    sort(v.begin(),v.end()); //    First Sort the vector/array if it's not already sorted because binary search can only be
    // implemented on monotonic function

    int low =0 , high = v.size()-1; // Search space is whole vector size initially

    while(high - low >1){
        int mid = low +( (high-low) >> 1);   // mid = (low +high)/2 in efficient way using right shift operator...
        if(v[mid]<element_toFind){
            low = mid+1;   // if element_toFind is greater than v[mid] then transfer the low to mid+1
        }else{
            high = mid;
        }


    }
    if(v[low] ==element_toFind ){
        return low;
    }else if(v[high] == element_toFind){
        return high;
    }else{
        return -1;
    }
}

int main(){

    vector<int> v = {2,3,4,7,8,9,12,14,16,17};

    cout<< binarySearch(v,15)<<endl;
    cout<< binarySearch(v,12)<<endl;
    cout<< binarySearch(v,2)<<endl;




    return 0;
}

File: JAVA/BinarySearch.java
/*
    Code: Binary Search Using Java(Iterative Approach)
    Name: Gaurav Agarwal
    Date of Modification: 2 oct 2022
*/
import java.util.Scanner;
public class BinarySearch{
/*
    Binary search is used to search a key element from multiple elements. Binary search is faster than linear search.

    In case of binary search, array elements must be in ascending order. If you have unsorted array, you can sort the array using Arrays.sort(arr) method.

*/
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
        System.out.println("Enter the size of the array");
        int n = sc.nextInt();

        // declaring array
        int arr[] = new int[n];
        System.out.println("Enter the elements:");

        // taking input array elements from user
        for(int i=0;i<n;i++)
            arr[i] = sc.nextInt();
        System.out.println("Enter the element to be searched:");

        // element to be searched
        int search = sc.nextInt();
        int low = 0;
        int high = n-1;
        int mid;
        while(low<=high)
        {
            mid = (high+low)/2;
            if(arr[mid]==search)
            {
                System.out.println(search+" found at "+mid+" index");
                break;
            }
            else if(arr[mid]<search)
                low = mid+1;
            else
                high = mid-1;
        }
        if(low>high)
            System.out.println("element not found in array.");

        /*
            Time complexity: O(logn)
            because if we do not find in middle then we will compare with the half of the remaining list.

            Space complexity: O(1)
            we are not using any extra array 
        */
	}

}
